core/arch/riscv/mm/core_mmu_arch.c

```diff
@@ -0,0 +1,717 @@
+{
+	switch (op) {
+	case DCACHE_CLEAN:
+		dcache_op_all(DCACHE_OP_CLEAN);
+		break;
+	case DCACHE_AREA_CLEAN:
+		dcache_clean_range(va, len);
+		break;
+	case DCACHE_INVALIDATE:
+		dcache_op_all(DCACHE_OP_INV);
+		break;
+	case DCACHE_AREA_INVALIDATE:
+		dcache_inv_range(va, len);
+		break;
+	case ICACHE_INVALIDATE:
+		icache_inv_all();
+		break;
+	case ICACHE_AREA_INVALIDATE:
+		icache_inv_range(va, len);
+		break;
+	case DCACHE_CLEAN_INV:
+		dcache_op_all(DCACHE_OP_CLEAN_INV);
+		break;
+	case DCACHE_AREA_CLEAN_INV:
+		dcache_cleaninv_range(va, len);
+		break;
+	default:
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+	return TEE_SUCCESS;
+}
+
+unsigned int asid_alloc(void)
+{
+	uint32_t exceptions = cpu_spin_lock_xsave(&g_asid_spinlock);
+	unsigned int r = 0;
+	int i = 0;
+
+	bit_ffc(&g_asid, BIT(g_asid), &i);
+	if (i == -1) {
+		r = 0;
+	} else {
+		bit_set(&g_asid, i);
+		r = i + 1;
+	}
+
+	cpu_spin_unlock_xrestore(&g_asid_spinlock, exceptions);
+
+	return r;
+}
+
+void asid_free(unsigned int asid)
+{
+	uint32_t exceptions = cpu_spin_lock_xsave(&g_asid_spinlock);
+
+	/* Only even ASIDs are supposed to be allocated */
+	assert(!(asid & 1));
+
+	if (asid) {
+		int i = asid - 1;
+
+		assert(i < BIT(g_asid) && bit_test(&g_asid, i));
+		bit_clear(&g_asid, i);
+	}
+
+	cpu_spin_unlock_xrestore(&g_asid_spinlock, exceptions);
+}
+
+bool arch_va2pa_helper(void *va, paddr_t *pa)
+{
+	vaddr_t vaddr = (vaddr_t)va;
+	struct mmu_pgt *pgt = NULL;
+	struct mmu_pte *pte = NULL;
+	int level = 0;
+	unsigned int idx = 0;
+	struct mmu_partition *prtn = core_mmu_get_prtn();
+
+	if (!pa)
+		return false;
+
+	pgt = core_mmu_get_root_pgt_va(prtn);
+
+	for (level = CORE_MMU_BASE_TABLE_LEVEL; level >= 0; level--) {
+		idx = core_mmu_pgt_idx(vaddr, level);
+		pte = core_mmu_table_get_entry(pgt, idx);
+
+		if (core_mmu_entry_is_invalid(pte)) {
+			return false;
+		} else if (core_mmu_entry_is_leaf(pte)) {
+			*pa = pte_to_pa(pte) |
+				(vaddr & (BIT64(RISCV_PGSHIFT) - 1));
+			return true;
+		}
+
+		pgt = phys_to_virt(pte_to_pa(pte),
+				   MEM_AREA_TEE_RAM_RW_DATA, sizeof(*pgt));
+	}
+
+	return false;
+}
+
+bool cpu_mmu_enabled(void)
+{
+	return read_satp();
+}
+
+bool core_mmu_find_table(struct mmu_partition *prtn, vaddr_t va,
+			 unsigned int max_level,
+			 struct core_mmu_table_info *tbl_info)
+{
+	struct mmu_pgt *pgt = NULL;
+	struct mmu_pte *pte = NULL;
+	unsigned int level = 0;
+	unsigned int idx = 0;
+
+	if (max_level >= RISCV_PGLEVELS)
+		return false;
+
+	if (!prtn)
+		prtn = core_mmu_get_prtn();
+
+	pgt = core_mmu_get_root_pgt_va(prtn);
+
+	for (level = CORE_MMU_BASE_TABLE_LEVEL; level > 0; level--) {
+		idx = core_mmu_pgt_idx(va, level);
+		pte = core_mmu_table_get_entry(pgt, idx);
+
+		if (core_mmu_entry_is_invalid(pte) ||
+		    core_mmu_entry_is_leaf(pte))
+			break;
+
+		pgt = phys_to_virt(pte_to_pa(pte),
+				   MEM_AREA_TEE_RAM_RW_DATA, sizeof(*pgt));
+	}
+
+	core_mmu_set_info_table(tbl_info, level, 0, pgt);
+
+	return true;
```

Comment:
I reference ARM's implementation which also considers the deepest level and `va_base` field.
Please take a look at it, thanks.

```suggestion
	while (true) {
		idx = core_mmu_pgt_idx(va - va_base, level);
		pte = core_mmu_table_get_entry(pgt, idx);

		if (level == deepest_level || level == 0 ||
		    core_mmu_entry_is_invalid(pte) ||
		    core_mmu_entry_is_leaf(pte)) {
			core_mmu_set_info_table(tbl_info, level, va_base, pgt);
			ret = true;
			goto out;
		}

		pgt = phys_to_virt(pte_to_pa(pte),
				   MEM_AREA_TEE_RAM_RW_DATA, sizeof(*pgt));
		if (!pgt)
			goto out;

		va_base += (vaddr_t)idx << CORE_MMU_SHIFT_OF_LEVEL(level);
		level--;
	}
out:
	return ret;
```

Reply:
@gagachang where is `va_base` defined ?